[InstCombine] Fix for folding select-like `shufflevector` into floating point binary operators.

michele-scandale opened the pull request:
Folding a select-like `shufflevector` into a floating point binary operators can only be done if the result is preserved for both case. In particular, if the common operand of the `shufflevector` and the floating point binary operator can be a NaN, then the transformation won't preserve the result value.

michele-scandale commented:
This should address the issue reported in https://github.com/llvm/llvm-project/issues/74326

goldsteinn commented on llvm/test/Transforms/InstCombine/shuffle_select.ll:
```diff
@@ -380,7 +380,7 @@ define <4 x double> @fdiv_constant_op0(<4 x double> %v) {
+define <4 x double> @fdiv_constant_op1(<4 x double> nofpclass(nan) %v) {
 ; CHECK-LABEL: @fdiv_constant_op1(
 ; CHECK-NEXT:    [[S:%.*]] = fdiv reassoc <4 x double> [[V:%.*]], <double undef, double 1.000000e+00, double 4.300000e+01, double 4.400000e+01>
 ; CHECK-NEXT:    ret <4 x double> [[S]]
```
Please keep the old tests so we can see the fix actually take affect.
You can add new tests with `nofpclass(nan)`.

michele-scandale replied:
I've added the test case `fadd_maybe_nan` to check that the transformation doesn't occur if the value might be a NaN. The other tests cases where I added `nofpclass(nan)` seems to cover the propagation of the fast-math flags in the transformation. Not sure if it is worth duplicating those test cases.

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -2159,6 +2160,18 @@ static Instruction *foldSelectShuffleWith1Binop(ShuffleVectorInst &Shuf) {
+  // This makes the transformation incorrect since the original program would
+  // have preserved the exact NaN bit-pattern.
+  // Avoid the folding if X can have NaN elements.
+  if (Shuf.getType()->isFPOrFPVectorTy() && !isKnownNeverNaN(X, 0, SQ))
```
Shut can't be a scalar FP type, so could just query the element type

michele-scandale committed:
[InstCombine] Fix for folding select-like `shufflevector` into floating point binary operators.

Folding a select-like `shufflevector` into a floating point binary
operators can only be done if the result is preserved for both case. In
particular, if the common operand of the `shufflevector` and the
floating point binary operator can be a NaN, then the transformation
won't preserve the result value.
